### ISSUE TITLE
Support Unit Testing of Userspace Programs

### ISSUE THREAD

**Issue by OliverGavin** [aya-rs/aya#999]
I have found myself writing mocks for different map types but it is a bit painful; when trying to use new map types I get similar type conflicts, etc until I realize I need to mock more things, IDE is no longer helpful for docs and autocomplete of methods not yet implemented by my mocks, etc.

Maybe there is a better way like creating my own wrapper types I can more easily mock. Ultimately I still end up having to write my own mock implementations and I think this is likely a problem faced by others so I wanted to open this issue to discuss it.

Example mock usage:
```rust
#[cfg(any(not(test), rust_analyzer))]
type PerCpuArray<V> = aya::maps::PerCpuArray<aya::maps::MapData, V>;

#[cfg(all(test, not(rust_analyzer)))]
use crate::mocks::aya::{Ebpf, PerCpuArray};
```

Example mock (quick and dirty):
```rust
#[derive(Clone)]
pub struct PerCpuArray<V: Pod> {
    inner: Arc<Mutex<Vec<Vec<V>>>>,
    _v: std::marker::PhantomData<V>,
}

impl<V: Pod> PerCpuArray<V> {
    pub(crate) fn new(len: usize, val: V) -> Self {
        let arr = vec![val.clone(); nr_cpus().unwrap()];
        PerCpuArray {
            inner: Arc::new(Mutex::new(vec![arr.clone(); len])),
            _v: core::marker::PhantomData,
        }
    }
}

impl TryFrom<Map> for PerCpuArray<u64> {
    type Error = MapError;

    fn try_from(_map: Map) -> Result<PerCpuArray<u64>, MapError> {
        Ok(PerCpuArray::new(1, 0u64))
    }
}

impl<V: Pod> PerCpuArray<V> {
    pub fn get(&self, index: &u32, _flags: u64) -> Result<PerCpuValues<V>, MapError> {
        let guard = self.inner.lock().unwrap();
        let values = guard.get(*index as usize).ok_or(MapError::OutOfBounds {
            index: *index,
            max_entries: guard.len() as u32,
        })?;
        PerCpuValues::try_from(values.clone()).map_err(|_| MapError::KeyNotFound)
    }

    pub fn set(&mut self, index: u32, values: PerCpuValues<V>, _flags: u64) -> Result<(), MapError> {
        let arr = (0..nr_cpus().unwrap())
            .map(|i| values.get(i).unwrap().clone())
            .collect::<Vec<V>>();
        let mut guard = self.inner.lock().unwrap();
        guard[index as usize] = arr;
        Ok(())
    }
}
```

Similar to https://github.com/aya-rs/aya/issues/36

----

One thing I did notice was that tests inside aya leverage `test_utils`: https://github.com/aya-rs/aya/blob/a167554b8b1b67ca329721381afae9992a3c1544/aya/src/maps/mod.rs#L1040-L1076

which uses `override_syscall`: https://github.com/aya-rs/aya/blob/a167554b8b1b67ca329721381afae9992a3c1544/aya/src/sys/fake.rs#L19

Which is used as a mock here: https://github.com/aya-rs/aya/blob/a167554b8b1b67ca329721381afae9992a3c1544/aya/src/sys/mod.rs#L91

However, this is only usable internally to the crate and when compiled to a test target. It would be really cool to expose something similar to make it easier to test for users of aya.

There are a few different ways around it but I am thinking the simplest and least intrusive could be a feature that users can enable for their tests that allows them to override syscalls. Going one step further, the creation of the maps themselves could be mocked appropriately.

**Comment by provokateurin** [comment]
@tamird this can be closed now

**Comment by tamird** [comment]
Fixed in #1461.

**Comment by OliverGavin** [comment]
Can we reopen this please? `BPF_PROG_TEST_RUN` helps running BPF in tests but I don't think it helps with mocking of user space map operations which perform real syscalls.

**Comment by tamird** [comment]
Ah, this is about testing userspace. I'm not sure how we will accomplish that without writing in-process implementations for all the map types, but OK, i can reopen.

**Comment by OliverGavin** [comment]
I think we can handle this in two steps; 
1) expose some of the internal override tooling you have already in your tests (easy)
2) build some helpers on top of that to make it less painful for users - but they can also make their own helpers using the same raw primitives

Or if you have other ideas like having some user space stub data structures that get used when a feature is enabled? That would largely eliminate the need for users to mock.

I'd be happy to take a stab at this if you would review a PR on it. Let me know which way you are leaning with it - either a light set of hooks or a more integrated testing experience. 

I've already made an attempt at 1) today and applied it to another project to test. I did start to realize that it's arguably more ugly that my current approach without any helpers as I need to wire up the data structures in my own tests. On the other hand it gives me an actual data structure that works. It can be a short term thing. Let me know if you want me to share that one or which direction you would prefer.

**Comment by OliverGavin** [comment]
@tamird I started a PR with the general idea in #1565 if you have any thoughts on it

Also, you can see here how it can be used to replace mocks I have:
https://github.com/aya-metrics-rs/aya-metrics/compare/main...feat/remove-mocks-use-test-utils

**Comment by vadorovsky** [comment]
> I think we can handle this in two steps;
> 
>     1. expose some of the internal override tooling you have already in your tests (easy)
> 
>     2. build some helpers on top of that to make it less painful for users - but they can also make their own helpers using the same raw primitives
> 

I agree with this direction. I think your PR is a good start to build more high-level mocks/helpers further on.

> Or if you have other ideas like having some user space stub data structures that get used when a feature is enabled? That would largely eliminate the need for users to mock.

We can think of it once the more low-level primitives are exposed.
